### PR TITLE
dbus-proxy: Improve filtering security

### PR DESCRIPTION
**Pull request by alivenets**
Add additional checks for D-Bus method "org.freedesktop.DBus.Hello" and
implementation-defined message "org.freedesktop.DBus.Local.Disconnected"

Make sricter check on forwarding, remove weak condition if interface is
"org.freedesktop.DBus"

Add additional checks for eavesdropping

Signed-off-by: Alexander Livenets <alivenets@zoho.eu>

### CODE REVIEW

**Comment by alivenets** [comment]
@martin-ejdestig please review

File: src/proxy.c [hunk 1]
```diff
@@ -98,13 +98,17 @@ DBusHandlerResult filter_cb (DBusConnection *conn,
...
+                    dbus_message_get_member(msg) != NULL      &&
                     strcmp (dbus_message_get_path(msg),
-                            "/org/freedesktop/DBus") == 0 &&
+                            "/org/freedesktop/DBus") == 0     &&
                     strcmp (dbus_message_get_interface(msg),
-                            "org.freedesktop.DBus")  == 0 &&
+                            "org.freedesktop.DBus")  == 0     &&
                     strcmp (dbus_message_get_destination(msg),
-                            "org.freedesktop.DBus")  == 0 &&
+                            "org.freedesktop.DBus")  == 0     &&
```
Comment (martin-ejdestig) [inline src/proxy.c]: Looks correct to me. All these getters are for header fields that are optional and libdbus API documentation mentions that they return NULL if field is not set.

Minor nit (feel free to leave as is): perhaps move NULL check next to corresponding `strcmp()` function?

File: src/proxy.c [hunk 2]
```diff
@@ -134,6 +138,8 @@ DBusHandlerResult filter_cb (DBusConnection *conn,
     /* Handle Disconnected */
     if (dbus_message_get_type(msg) ==
                 DBUS_MESSAGE_TYPE_SIGNAL                   &&
+                dbus_message_get_interface(msg) != NULL    &&
+                dbus_message_get_member(msg) != NULL       &&
```
Comment (martin-ejdestig) [inline src/proxy.c]: Same as above. (Perhaps just move so corresponding null check directly preceeds strcmp() where it is used.)

File: src/proxy.c [hunk 3]
```diff
@@ -152,18 +158,10 @@ DBusHandlerResult filter_cb (DBusConnection *conn,
...
-                        msg,
-                        &serial);
-    } else if (is_allowed("outgoing",
-                          dbus_message_get_interface (msg),
-                          dbus_message_get_path      (msg),
-                          dbus_message_get_member    (msg)))
+    if (is_allowed("outgoing",
+                   dbus_message_get_interface (msg),
+                   dbus_message_get_path      (msg),
+                   dbus_message_get_member    (msg)))
```
Comment (martin-ejdestig) [inline src/proxy.c:164]: This I am a bit more unsure of. Why not always allow the "org.freedesktop.DBus" interface? It is for talking to the message bus itself. More or less all configurations will have to have this in them if this is removed?

File: src/proxy.c [hunk 4]
```diff
@@ -477,7 +475,8 @@ gboolean is_incoming_eavesdropping (DBusMessage *msg)
                                DBUS_TYPE_STRING,
                                &msg_arguments);
 
-        if (strstr(msg_arguments, "eavesdrop=true") != NULL)
+        if (strstr(msg_arguments, "eavesdrop=true") != NULL ||
+            strstr(msg_arguments, "eavesdrop='true'") != NULL)
```
Comment (martin-ejdestig) [inline src/proxy.c:479]: Do you know if `eavesdrop=true` is valid? It looks like https://dbus.freedesktop.org/doc/dbus-specification.html only mentions `eavesdrop='true'` but perhaps I am missing some "escape rule".

But also, no harm in leaving the string without `'` in it... I guess.

**Review by martin-ejdestig — CHANGES_REQUESTED** [review]
Only "serious" thing I am wondering about is the removal of always allowing the "org.freedesktop.DBus" interface.

File: src/proxy.c (hunk 4 above)
Comment (alivenets) [reply src/proxy.c:479]: According to match rule parsing code, function doing that: https://github.com/freedesktop/dbus/blob/master/bus/signals.c, both tokens which are to the right side of `=` are valid: `true` and `'true'`

File: src/proxy.c (hunk 3 above)
Comment (alivenets) [reply src/proxy.c:164]: `org.free.desktop.DBus` interface can provide extensive information to clients in containers, which can be dangerous. For example, all objects on bus and implemented interfaces. So, there should be granular access control.

File: src/proxy.c (hunk 1 above)
Comment (alivenets) [reply src/proxy.c]: Reorganized

File: src/proxy.c (hunk 2 above)
Comment (alivenets) [reply src/proxy.c]: Reorganized

File: src/proxy.c (hunk 4 above)
Comment (martin-ejdestig) [reply src/proxy.c:479]: You are correct. :) https://github.com/freedesktop/dbus/blob/03158095968de8a9542b7ce448d5e9b36f10d827/bus/signals.c#L591 removes `\'` so both `strstrs()` should be kept.

File: src/proxy.c (hunk 3 above)
Comment (martin-ejdestig) [reply src/proxy.c:164]: Ah, information that was missing for me here was what dbus-proxy is mainly being used for. If it is for filtering in containers it probably makes sense to be as restrictive as possible by default and require configuration to let things through.

**Comment by martin-ejdestig** [comment]
It initially looked to me like there would be similar NULL-checks needed for  `dbus_message_get*_()` when passed to `is_allowed()`. But strings are eventually passed to `g_pattern_match_simple()` that handles NULL, so I think we are OK there. :)

**Comment by martin-ejdestig** [comment]
There is a `dbus_message_get_interface(msg) == NULL` check in `master_filter_cb()` that looks like it should be inverted though.

```C
    /* Forward */
    if (dbus_message_get_interface(msg) == NULL ||
        strcmp(dbus_message_get_interface(msg),
               "org.freedesktop.DBus")  == 0)
    {
        if (is_incoming_eavesdropping(msg) &&
```

If there is no interface set in the header, it does not make sense to check if it is an eavesdropping message. Would you mind taking a look at that? (Just thinking that it is potential issue for project you are using dbus-proxy in similar to what you initially fixed.)

I also wonder a bit about `is_allowed("incoming")` not always being called in `master_filter_cb()`, similar to what is done in `filter_cb()` after your changes.

Anyway, OK if we do not take those in this PR. (Just say so and we will merge this PR. :) Just things I wonder about now looking at the rest of the code for more or less the first time.

Would you mind squashing the "style" fixup commit into the first commit? (I am all for multiple small commits in PR:s, but might as well squash away "fixups" that are not interesting for posterity. :)

**Comment by alivenets** [comment]
Regarding the condition in `master_filter_cb()` - I will try to find the case of the condition violation. Therefore, I suggest to merge this PR first, and then other may come with the fix of your proposal.

**Comment by martin-ejdestig** [comment]
Sounds good.

Since we do not have any CI setup for dbus-proxy I manually verified that tests still pass by running:

```shell
$ rm -rf build && cmake -H. -Bbuild && make -j8 -C build && (cd component-test/; py.test -v -s)`
```

**Pull request merged**